p2p/trust/trustmetric.go

```diff
@@ -0,0 +1,574 @@
+	IntervalLength time.Duration
+}
+
+// DefaultConfig returns a config with values that have been tested and produce desirable results
+func DefaultConfig() TrustMetricConfig {
+	return TrustMetricConfig{
+		ProportionalWeight: 0.4,
+		IntegralWeight:     0.6,
+		TrackingWindow:     (time.Minute * 60 * 24) * 14, // 14 days.
+		IntervalLength:     1 * time.Minute,
+	}
+}
+
+// NewMetric returns a trust metric with the default configuration
+func NewMetric() *TrustMetric {
+	return NewMetricWithConfig(DefaultConfig())
+}
+
+// NewMetricWithConfig returns a trust metric with a custom configuration
+func NewMetricWithConfig(tmc TrustMetricConfig) *TrustMetric {
+	tm := new(TrustMetric)
+	config := customConfig(tmc)
+
+	// Setup using the configuration values
+	tm.proportionalWeight = config.ProportionalWeight
+	tm.integralWeight = config.IntegralWeight
+	tm.intervalLen = config.IntervalLength
+	// The maximum number of time intervals is the tracking window / interval length
+	tm.maxIntervals = int(config.TrackingWindow / tm.intervalLen)
+	// The history size will be determined by the maximum number of time intervals
+	tm.historyMaxSize = intervalToHistoryOffset(tm.maxIntervals) + 1
+	// This metric has a perfect history so far
+	tm.historyValue = 1.0
+	// Setup the channels
+	tm.update = make(chan *updateBadGood, defaultUpdateChanCapacity)
+	tm.trustValue = make(chan *reqTrustValue, defaultRequestChanCapacity)
+	tm.stop = make(chan bool, 2)
+
+	go tm.processRequests()
+	return tm
+}
+
+/* Private methods */
+
+// Ensures that all configuration elements have valid values
+func customConfig(tmc TrustMetricConfig) TrustMetricConfig {
+	config := DefaultConfig()
+
+	// Check the config for set values, and setup appropriately
+	if tmc.ProportionalWeight > 0 {
+		config.ProportionalWeight = tmc.ProportionalWeight
+	}
+
+	if tmc.IntegralWeight > 0 {
+		config.IntegralWeight = tmc.IntegralWeight
+	}
+
+	if tmc.IntervalLength > time.Duration(0) {
+		config.IntervalLength = tmc.IntervalLength
+	}
+
+	if tmc.TrackingWindow > time.Duration(0) &&
+		tmc.TrackingWindow >= config.IntervalLength {
+		config.TrackingWindow = tmc.TrackingWindow
+	}
+
+	return config
+}
+
+// Calculates the derivative component
+func (tm *TrustMetric) derivativeValue() float64 {
+	return tm.proportionalValue() - tm.historyValue
+}
+
+// Strengthens the derivative component when the change is negative
+func (tm *TrustMetric) weightedDerivative() float64 {
+	var weight float64
+
+	d := tm.derivativeValue()
+
+	if d < 0 {
+		weight = 1.0
+	}
+	return weight * d
+}
+
+// Performs the update for our Faded Memories process, which allows the
+// trust metric tracking window to be large while maintaining a small
+// number of history data values
+func (tm *TrustMetric) updateFadedMemory() {
+	if tm.historySize < 2 {
+		return
+	}
+
+	first := tm.historySize - 1
+	// Keep the most recent history element
+	for count, i := 1, first-1; count < tm.historySize; count, i = count+1, i-1 {
```

Comment:
this is hard to read. is it possible to simplify?

Reply:
Sure